### PR TITLE
[RSDK-17] Delayed push

### PR DESCRIPTION
**Pull request by aagorelik**
* added cache for data points
* remapped add_data_point to write to local storage
* ruby library version number is sent with every request

Issues: RSDK-5, RSDK-20, RSDK-17

### CODE REVIEW

File: lib/vchain/cache/sqlite.rb [hunk 1]
```diff
@@ -134,6 +159,146 @@ def update_last_seen_block_for_address(address, max_block_height)
...
+                   record.push_timeout, 
+                   record.push_operation_id,
+                   record.uuid
+                  ])
+
+      record
+    end
+
+    def lock_and_update(params, filter)
+      if params.length == 0 || filter.length == 0
```
Comment (baitcode) [inline lib/vchain/cache/sqlite.rb]: don't see a lock here

File: lib/vchain/cache/sqlite.rb [hunk 2]
```diff
@@ -134,6 +159,146 @@ def update_last_seen_block_for_address(address, max_block_height)
       end
     end
 
+    def add_data_points_record(record)
```
Comment (baitcode) [inline lib/vchain/cache/sqlite.rb]: check out INSERT OR REPLACE statement

File: lib/vchain/cache/sqlite.rb [hunk 3]
```diff
@@ -2,6 +2,8 @@ module VChain::Cache
 
   class SqliteAdapter < VChain::Cache::Adapter
 
+    @@sync_mutex = Mutex.new
```
Comment (baitcode) [inline lib/vchain/cache/sqlite.rb]: I can't find in doc what happens if thread stumbles into synchronize block that is already locked. Is it going to raise exception?

Comment (aagorelik) [inline lib/vchain/cache/sqlite.rb]: It is going to wait until lock is gone

**Pull request merged**